plasmapy/utils/data/downloader.py

```diff
@@ -4,74 +4,409 @@
+
+        try:
+            reply = requests.get(url)  # noqa: S113
+
+        # No test coverage for this exception since we can't test it without
+        # severing the network connectivity in pytest
+        except requests.ConnectionError as err:  # coverage: ignore
+            raise requests.ConnectionError(
+                "Unable to connect to data " f"repository {self._API_BASE_URL}"
+            ) from err
 
-        with path.open(mode="wb") as f:
+        # Extract the 'message' value if it is there
+        # If the file does not exist on the repository, the GitHub API
+        # will return `Not Found` in response to this but not raise a 404 error
+        if reply.status_code == 404:
+            raise ValueError(f"URL returned 404: {url}")
+
+        return reply
+
+    def _filepath(self, filename: str) -> Path:
+        """Formats a filepath from a filename."""
+        return Path(self._download_directory, filename)
+
+    def _download_file(self, filename: str, dl_url: str) -> Path:
+        """
+        Download a file from a given URL to a specified path.
+
+        Parameters
+        ----------
+        filename : str
+            Name of the file to download: determines the filepath
+
+        dl_url : str
+            URL from which to download
+
+
+        Returns
+        -------
+        filepath : str
+            Path to the downloaded file
+
+        """
+
+        # Request the contents of the file from the download URL
+        reply = self._http_request(dl_url)
+
+        filepath = self._filepath(filename)
+
+        # Write the contents to file
+        with filepath.open(mode="wb") as f:
             f.write(reply.content)
 
-    return path
+        return filepath
+
+    def _get_file_without_validation(self, filename: str) -> Path:
+        """
+        Return file logic without validation.
+
+        Returns
+        -------
+        filepath : Path
+            Path to file
+
+        Raises
+        ------
+        ValueError
+            If the resource cannot be found locally or on the repository
+
+        """
+        filepath = self._filepath(filename)
+
+        # If the file exists locally, return that
+        if filepath.is_file():
+            return filepath
+
+        # Try blindly downloading from the base URL
+        # Note that downloading directly from the RAW url does not
+        # require an API call.
+        try:
+            dl_url = urljoin(self._RAW_BASE_URL, filename)
+            return self._download_file(filename, dl_url)
+        except ValueError:
+            pass
+
+        raise ValueError(
+            "Resource could not be found locally or "
+            "retrieved from the PlasmPy-data repository: "
+            f"{filename}."
+        )
+
+    def _get_file_with_validation(self, filename: str) -> Path:
+        """
+        Return file logic with validation.
+        """
+        filepath = self._filepath(filename)
+
+        # Try to update the repo blob dict
```

Comment:
```suggestion
```
We can probably avoid this comment since it's very similar to the name of the method being called.

A corollary here is: good choice for the method name!